initial-promises/node-server/common/1.0/update.cf

```diff
@@ -186,155 +237,198 @@ bundle agent update_action
-        classes           => success("rudder_ncf_local_updated", "rudder_ncf_local_update_error", "rudder_ncf_local_updated_ok"),
-        comment           => "Update the local Rudder ncf instance";
+        classes           => success("rudder_ncf_hash_update_repaired", "rudder_ncf_hash_update_error", "rudder_ncf_hash_update_ok");
 
-      "${client_inputs}/${file_to_check_update}"
-        copy_from         => remote_unsecured_without_perms("${server_info.cfserved}","${server_inputs}/${file_to_check_update}"),
+      "${g.rudder_ncf}/local/${g.rudder_ncf_hash_file}"
+        copy_from         => remote_unsecured_without_perms("${server_info.cfserved}","${g.rudder_ncf_origin_local_hash}"),
         action            => immediate,
         move_obstructions => "true",
-        classes           => success("rudder_promises_generated_repaired", "rudder_promises_generated_error", "rudder_promises_generated_ok");
+        classes           => success("rudder_ncf_hash_update_repaired", "rudder_ncf_hash_update_error", "rudder_ncf_hash_update_ok");
 
     # The defined class are persistent, so if they are already set, promises has already been updated
     # a short while ago
-    rudder_promises_generated_repaired.!root_server::
+    new_promises_available.!root_server::
       "${client_inputs}"
         copy_from         => remote("${server_info.cfserved}","${server_inputs}"),
         depth_search      => recurse("inf"),
         action            => immediate,
         move_obstructions => "true",
+        file_select       => exclude("${rudder_promises_timestamp_tmp}"),
         classes           => success("config", "no_update", "config_ok");
 
+    # We copy ncf on non root server if ncf hash file changed, or if  promises have been updated
+    !root_server.(new_promises_available|rudder_ncf_hash_update_repaired)::
+      "${g.rudder_ncf}/common"
+        copy_from         => remote_unsecured_without_perms("${server_info.cfserved}", "${g.rudder_ncf_origin_common}"),
+        depth_search      => recurse_ignore("inf", @{g.ncf_common_excludedreps}),
+        perms             => u_mog("644", "root", "0"),
+        action            => immediate,
+        move_obstructions => "true",
+        classes           => success("rudder_ncf_common_updated", "rudder_ncf_common_update_error", "rudder_ncf_common_updated_ok"),
+        comment           => "Update the common Rudder ncf instance";
+
+      "${g.rudder_ncf}/local"
+        copy_from         => remote("${server_info.cfserved}", "${g.rudder_ncf_origin_local}"),
+        depth_search      => recurse_ignore("inf", @{g.excludedreps}),
+        perms             => u_mog("644", "root", "0"),
+        action            => immediate,
+        move_obstructions => "true",
+        classes           => success("rudder_ncf_local_updated", "rudder_ncf_local_update_error", "rudder_ncf_local_updated_ok"),
+        comment           => "Update the local Rudder ncf instance";
```

Comment:
This means we will copy/update ncf with initial-promises, therefore at times when the node this is running on has not been accepted. In all likelihood, this means there will be times when we can update ncf (since the cf-serverd ACLs allow access to ncf for all nodes in the allowed networks - except for local/50_techniques). If there is a significant change in ncf and the initial promises use ncf, this could break initial promises, and never get fixed.

I think this needs some thought.